src/Shared/Model/PackageExistence.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Model;
+
+using Enums;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+public record PackageExistence
+{
+    public PackageExistence(bool exists, bool didExist, IEnumerable<PackageDoesNotExistReason>? doesNotExistReasons = null)
+    {
+        if (exists && !didExist)
+        {
+            throw new InvalidOperationException("The package can't have Exists as true, and DidExist as false!");
+        }
+
+        this.Exists = exists;
+        this.DidExist = didExist;
+        this.DoesNotExistReasons = doesNotExistReasons ?? new List<PackageDoesNotExistReason>();
+
+        if (this.DoesNotExistReasons.Contains(PackageDoesNotExistReason.VersionUnpublished))
+        {
+            throw new InvalidOperationException("The package can't have VersionUnpublished as a does not exist reason!");
+        }
+    }
+
+    public IEnumerable<PackageDoesNotExistReason> DoesNotExistReasons { get; init; }
+    public bool DidExist { get; init; }
+    public bool Exists { get; init; }
+}
+
+public record PackageVersionExistence
+{
+    public PackageVersionExistence(bool exists, bool didExist, bool packageExists, bool packageDidExist, IEnumerable<PackageDoesNotExistReason>? doesNotExistReasons = null)
+    {
+        if (exists && !didExist)
+        {
+            throw new InvalidOperationException("The package version can't have Exists as true, and DidExist as false!");
+        }
+        
+        if (packageExists && !packageDidExist)
+        {
+            throw new InvalidOperationException("The package version can't have PackageExists as true, and PackageDidExist as false!");
+        }
+        
+        if ((exists && !packageExists) || (didExist && !packageDidExist))
+        {
+            throw new InvalidOperationException("The package version can't exist or have existed, if the package doesn't or never did exist!");
+        }
+
+        this.Exists = exists;
+        this.DidExist = didExist;
+        this.PackageExists = packageExists;
+        this.PackageDidExist = packageDidExist;
+        this.DoesNotExistReasons = doesNotExistReasons ?? new List<PackageDoesNotExistReason>();
+    }
+
+    public IEnumerable<PackageDoesNotExistReason> DoesNotExistReasons { get; init; }
+    public bool DidExist { get; init; }
+    public bool Exists { get; init; }
+    public bool PackageDidExist { get; init; }
+    public bool PackageExists { get; init; }
+}
```

Comment:
This feels a bit complex to me. Here's an alternative which looks a bit simpler to my eyes (the below is just for package version - we would do the same for package as well:

```c#
using System.Collections.Generic;

public enum PackageVersionRemovalReason
{
    Unpublished,
    ConsideredMalicious,
}

interface IPackageVersionExistence
{
    bool Exists { get; }
    bool HasEverExisted { get; }
    bool WasRemoved => !this.Exists && this.HasEverExisted;
}

public record PackageVersionExists : IPackageVersionExistence
{
    public bool Exists => true;
    public bool HasEverExisted => true;
}

public record PackageVersionNotFound : IPackageVersionExistence
{
    public bool Exists => false;
    public bool HasEverExisted => false;
}

public record PackageVersionRemoved(IReadOnlySet<PackageVersionRemovalReason> RemovalReasons) : IPackageVersionExistence
{
    public bool Exists => false;
    public bool HasEverExisted => true;
}
```

What do you think? 

One of the advantages of the above format is that it lends itself to [pattern matching by type](https://endjin.com/blog/2022/02/pattern-matching-in-csharp) rather nicely.

Reply:
Addressed!